[Bug] Can't use micronaut-data-jdbc with @PostConstruct and eagerInitSingletons

Thanks for reporting an issue, please review the task list below before submitting the
issue. Your issue report will be closed if the issue is incomplete and the below tasks not completed.

NOTE: If you are unsure about something and the issue is more of a question a better place to ask questions is on Stack Overflow (https://stackoverflow.com/tags/micronaut) or Gitter (https://gitter.im/micronautfw/). DO NOT use the issue tracker to ask questions.

### Task List

- [] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [ ] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

I decided to play with the micronaut's settings and try different modes in my small service. There was a problem with the selection `eagerInitSingletons` mode.

In one of my service classes, there is data loading from the database after the bean is created: 
```java
public class SimpleService {

    @PostConstruct
    public void init() {
        loadFromDb();
        checkState();
    }
}
```

My Application:
```java
        Micronaut.build(args)
                .classes(Application.class)
                .banner(false)
                .eagerInitSingletons(true)
                .start();
```

Ok, when I try to start application I get the exception:
```
Caused by: io.micronaut.data.exceptions.DataAccessException: Cannot convert type [class java.sql.Timestamp] with value [2021-04-12 14:58:10.85808] to target type: LocalDateTime updateTime. Consider defining a TypeConverter bean to handle this case.
	at io.micronaut.data.runtime.mapper.ResultReader.lambda$convertRequired$1(ResultReader.java:68)
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at io.micronaut.data.runtime.mapper.ResultReader.convertRequired(ResultReader.java:67)
	at io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper.convertAndSet(SqlResultEntityTypeMapper.java:648)
	at io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper.getJoins(SqlResultEntityTypeMapper.java:483)
	at io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper.readEntity(SqlResultEntityTypeMapper.java:359)
	at io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper.map(SqlResultEntityTypeMapper.java:168)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations$2.tryAdvance(DefaultJdbcRepositoryOperations.java:379)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:326)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.lambda$findAll$7(DefaultJdbcRepositoryOperations.java:452)
	at io.micronaut.transaction.support.AbstractSynchronousTransactionManager.executeRead(AbstractSynchronousTransactionManager.java:155)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.findAll(DefaultJdbcRepositoryOperations.java:450)
	at io.micronaut.data.runtime.intercept.DefaultFindAllInterceptor.intercept(DefaultFindAllInterceptor.java:52)
	at io.micronaut.data.runtime.intercept.DefaultFindAllInterceptor.intercept(DefaultFindAllInterceptor.java:36)
	at io.micronaut.data.intercept.DataIntroductionAdvice.intercept(DataIntroductionAdvice.java:80)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:96)
	at io.micronaut.validation.ValidatingInterceptor.intercept(ValidatingInterceptor.java:137)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:96)
```

I want to say that this error does not exist with the lazy mode of creating beans. 

### Environment Information

- **Operating System:** Windows 10
- **Micronaut Version:** 2.4.2
- **JDK Version:** zulu 11.0.10

### Example Application

- TODO: link to github repository with example that reproduces the issue

--- a/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2EagerContextSpec.groovy
+++ b/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2EagerContextSpec.groovy
@@ -1,0 +1,66 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.jdbc.h2
+
+import io.micronaut.context.ApplicationContext
+import io.micronaut.context.BeanLocator
+import io.micronaut.data.jdbc.config.SchemaGenerator
+import io.micronaut.data.tck.entities.Person
+import jakarta.annotation.PostConstruct
+import jakarta.inject.Singleton
+import spock.lang.AutoCleanup
+import spock.lang.Shared
+import spock.lang.Specification
+
+class H2EagerContextSpec extends Specification implements H2TestPropertyProvider {
+
+    @AutoCleanup
+    @Shared
+    ApplicationContext context = ApplicationContext.builder(getProperties() + [
+            'eager-test'            : true,
+            'datasources.other.name': 'other'
+    ])
+            .eagerInitSingletons(true)
+            .build()
+            .start()
+
+    void "test eager start"() {
+        when:
+            def personRepository = context.getBean(H2PersonRepository)
+        then:
+            personRepository.findAll().toList().size() == 4
+    }
+
+    @Singleton
+    static class SimpleService {
+
+        private final H2PersonRepository personRepository
+
+        SimpleService(H2PersonRepository personRepository) {
+            this.personRepository = personRepository;
+        }
+
+        @PostConstruct
+        void init(SchemaGenerator schemaGenerator, BeanLocator beanLocator) {
+            schemaGenerator.createSchema(beanLocator)
+
+            personRepository.save(new Person(name: 'a'))
+            personRepository.save(new Person(name: 'c'))
+            personRepository.save(new Person(name: 'b'))
+            personRepository.save(new Person(name: 'd'))
+        }
+    }
+}

